macro Operator docs (#4767)
SumoLogic/sumologic-documentation: docs/search/search-query-language/search-operators/macro.md

---
id: macro
title: macro Operator
sidebar_label: macro
---
import useBaseUrl from '@docusaurus/useBaseUrl';

> With `macro` operator you can reuse the portions of search query language in multiple other queries. This operator also includes arguments and also performs its respective evaluation of the arguments.

Reviewer: ```suggestion
The `macro` operator allows you to reuse portions of a search query across multiple queries. Additionally, this operator includes arguments and performs its respective evaluation of the arguments.
```

> Use backquote (``) character to refer macro operator in the query. Macros referred inside another macro is called **Nested macros**. Macro referred within the context of another macro query of is called **Inner Macro**. The macro where the inner macro is referred is called the **Outer Macro**.

Reviewer: ```suggestion
To use `macro` the operator, reference it in your query with backticks (``). Macros can also be nested, enabling complex query reuse:
* **Nested Macros**. A macro referenced inside another macro.
* **Inner Macro**. A macro used within the context of another macro query.
* **Outer Macro**. The macro that references an inner macro.
```

:::note

> - Only **Administrators** and **Users** with access to **Query Reference** will be able to run queries using macros.

Reviewer: ```suggestion
- Only **Administrators** and **Users** with access to **Query Reference** can run queries using macros.
```

Reviewer 2: This should be **Macros** not query reference?

Author: I got this information from the macro operator confluence page. Please let me know if it is right to add it here?
![Uploading Screenshot 2024-11-28 at 11.22.01 AM.png…]()

> - Only users with **Administrator** access can create or delete the macro. 

Reviewer: ```suggestion
- Only users with **Administrator** access can create or delete macros.
```

Reviewer 2: We should remove `delete` here it is not a feature we support unless customer asks for it

:::

## Syntax

```
`<macro name>`
```

## Add a macro

> To create a macro follow the steps below:

Reviewer: ```suggestion
To create a macro, follow the steps below:
```

1. [**Classic UI**](/docs/get-started/sumo-logic-ui-classic). In the main Sumo Logic menu, select **Manage Data** > **Logs** > **Macros**. <br/> [**New UI**](/docs/get-started/sumo-logic-ui/). In the top menu, select **Manage Date**, and then under **Logs**, select **Macros**. You can also click the **Go To...** menu at the top of the screen and select **Macros**.
1. Click **+ Add Macro**.<br/><img src={useBaseUrl('img/search/searchquerylanguage/search-operators/macro-logs-page.png')} alt="macro-logs-page" style={{border: '1px solid gray'}} width="800" />
1. Or, in the log search page, select the part of search query language that needs to be reused and click on **Create Macro**.<br/><img src={useBaseUrl('img/search/searchquerylanguage/search-operators/macro-search-page.png')} alt="macro-search-page" style={{border: '1px solid gray'}} width="800" />
1. **Macro Details**. Enter the name for the macro. Description is optional.
1. **Macro Definition**. Enter the definition for the macro. To add arguments use the `{{Arg}}` syntax or select a part of the definition and click on **Add Argument**.

> 1. (Optional)**Arguments**. Enter the name and select the data type for the argument selected.
> 1. (Optional)**Argument Validation**. Define the validation condition and enter the error message that needs to be shown when the validation expression returns false.

Reviewer: ```suggestion
1. (Optional) **Arguments**. Enter the name and select the data type for the argument selected.
1. (Optional) **Argument Validation**. Define the validation condition and enter the error message that needs to be shown when the validation expression returns false.
```

1. **Usage**. Preview of how you use the macro in the log search.
1. Click **Submit** to save the macro.

### Limitations

- You can create a maximum of 50 macros.

> - You can add a maximum of 5 definitions.

Reviewer 2: We should add limit of 5 argument validations here too

> - You can add a maximum of 5 definitions.

Reviewer 2: This should be arguments, not definitions

### Example

> Consider the below query that search for the errors with timeslice of 5 minutes. 

Reviewer: ```suggestion
Consider the below query, which searches for errors with timeslice of 5 minutes. 
```

```
_sourceCategory=error | timeslice 5m
| count by _timeslice
```

> Now by creating macro for the timeslice field, the query with the `macro` operator is modified as:

Reviewer: ```suggestion
By creating a macro for the `timeslice` field, the query using the `macro` operator can be simplified as follows:
```

```

> _sourceCategory=error | `timeslice_macro`

Reviewer 2: This is correct

> _sourceCategory=error | `timeslice_macro`

Reviewer 2: We should add example of macro with arguments usage

---
id: macro
title: macro Operator
sidebar_label: macro
---
import useBaseUrl from '@docusaurus/useBaseUrl';

The `macro` operator allows you to reuse portions of a search query across multiple queries. Additionally, this operator includes arguments and performs its respective evaluation of the arguments.

To use `macro` the operator, reference it in your query with backticks (``). Macros can also be nested, enabling complex query reuse:
* **Nested Macros**. A macro referenced inside another macro.
* **Inner Macro**. A macro used within the context of another macro query.
* **Outer Macro**. The macro that references an inner macro.

:::note
- Only **Administrators** and **Users** with access to **Query Reference** will be able to run queries using macros.

> - Only users with **Administrator** access can create or delete macros.

Author: ```suggestion
- Only users with **Administrator** access can create macros.
```

1. Click **+ Add Macro**.<br/><img src={useBaseUrl('img/search/searchquerylanguage/search-operators/macro-logs-page.png')} alt="macro-logs-page" style={{border: '1px solid gray'}} width="800" />
1. Or, in the log search page, select the part of search query language that needs to be reused and click on **Create Macro**.<br/><img src={useBaseUrl('img/search/searchquerylanguage/search-operators/macro-search-page.png')} alt="macro-search-page" style={{border: '1px solid gray'}} width="800" />
1. **Macro Details**. Enter the name for the macro. Description is optional.
1. **Macro Definition**. Enter the definition for the macro. To add arguments use the `{{Arg}}` syntax or select a part of the definition and click on **Add Argument**.
1. (Optional) **Arguments**. Enter the name and select the data type for the argument selected.
1. (Optional) **Argument Validation**. Define the validation condition and enter the error message that needs to be shown when the validation expression returns false.
1. **Usage**. Preview of how you use the macro in the log search.
1. Click **Submit** to save the macro.

> 

Author: Added the "View and use the macro operator" section to address this comment.

### Limitations

- You can create a maximum of 50 macros.
- You can add a maximum of 5 arguments.

> 

Reviewer 2: We should put macros cannot be deleted as well. It is not an officially supported feature and will only be allowed under customer request.
